lib/DxilDia/DxcPixDxilStorage.cpp

```diff
@@ -0,0 +1,291 @@
+///////////////////////////////////////////////////////////////////////////////
+//                                                                           //
+// DxcPixDxilStorage.cpp                                                     //
+// Copyright (C) Microsoft Corporation. All rights reserved.                 //
+// This file is distributed under the University of Illinois Open Source     //
+// License. See LICENSE.TXT for details.                                     //
+//                                                                           //
+// Defines the DxcPixDxilStorage API.                                        //
+//                                                                           //
+///////////////////////////////////////////////////////////////////////////////
+
+#include "DxcPixDxilStorage.h"
+
+#include "dxc/Support/WinIncludes.h"
+#include "dxc/DxilPIXPasses/DxilPIXVirtualRegisters.h"
+#include "llvm/IR/Instructions.h"
+
+#include "DxcPixBase.h"
+#include "DxcPixLiveVariables.h"
+#include "DxcPixTypes.h"
+
+static HRESULT UnAliasType(
+    IDxcPixType *MaybeAlias,
+    IDxcPixType **OriginalType
+)
+{
+  CComPtr<IDxcPixType> Tmp(MaybeAlias);
+
+  HRESULT hr;
+  for (hr = Tmp->UnAlias(OriginalType);
+       hr == S_OK;
+       hr = Tmp->UnAlias(OriginalType))
+  {
+    Tmp.Attach(*OriginalType);
+  }
+
+  IFR(hr);
+  *OriginalType = Tmp.Detach();
+  return S_OK;
+}
+
+HRESULT CreateDxcPixStorageImpl(
+    dxil_debug_info::DxcPixDxilDebugInfo *pDxilDebugInfo,
+    IDxcPixType* OriginalType,
+    const dxil_debug_info::VariableInfo* VarInfo,
+    unsigned CurrentOffsetInBits,
+    IDxcPixDxilStorage** ppStorage)
+{
+  CComPtr<IDxcPixArrayType> ArrayTy;
+  CComPtr<IDxcPixStructType> StructTy;
+  CComPtr<IDxcPixScalarType> ScalarTy;
+
+  CComPtr<IDxcPixType> UnalisedType;
+  IFR(UnAliasType(OriginalType, &UnalisedType));
+
+  if (UnalisedType->QueryInterface(&ArrayTy) == S_OK)
+  {
+    return dxil_debug_info::NewDxcPixDxilDebugInfoObjectOrThrow<dxil_debug_info::DxcPixDxilArrayStorage>(
```

Comment:
The method has "OrThrow" in its name. Doesn't look like any callers catch before returning from a STDMETHOD, so maybe an exception boundary somewhere (ideally in the STDMETHOD)?

Reply:
Because there could be instances when the variable is not assigned to a register... Maybe. I'm leaving this here as a placeholder.

Reply:
All catches happen in DxcPixEntrypoints.cpp, as well as all default initialization for outparams